plugins/swf-backend/src/service/OpenApiService.ts

```diff
@@ -0,0 +1,156 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger } from 'winston';
+import { DiscoveryApi } from '@backstage/core-plugin-api';
+import fs from 'fs-extra';
+import { resolvePackagePath } from '@backstage/backend-common';
+
+export class OpenApiService {
+  logger: Logger;
+  discovery: DiscoveryApi;
+
+  constructor(logger: Logger, discovery: DiscoveryApi) {
+    this.logger = logger;
+    this.discovery = discovery;
+  }
+
+  private async fetchScaffolderActions(): Promise<any> {
+    const scaffolderUrl = await this.discovery.getBaseUrl('scaffolder');
+    return fetch(`${scaffolderUrl}/v2/actions`).then(res => {
+      return res.json();
+    });
+  }
+
+  private async openApiTemplate(): Promise<any> {
+    const path = resolvePackagePath(
+      `@backstage/plugin-swf-backend`,
+      `src/service/openapi-template.json`,
+    );
+    return fs.readFile(path).then(buffer => {
+      return JSON.parse(buffer.toString('utf8'));
+    });
+  }
+
+  async generateOpenApi(): Promise<any> {
+    const template = await this.openApiTemplate();
+    return this.fetchScaffolderActions()
+      .then(async actions => {
+        template.paths = await this.mapPaths(actions);
+        return actions;
+      })
+      .then(this.mapSchemas)
+      .then(schemas => {
+        template.components.schemas = schemas;
+        return template;
+      })
+      .catch(err => {
+        this.logger.error(err);
+      });
+  }
+
+  private async mapPaths(actions: any): Promise<any> {
+    const paths = {};
+    for (const action of actions) {
+      const actionId: string = action.id;
+      const description = action.description;
+      const schemaName = actionId.replaceAll(':', '');
+
+      const path = `/actions/${actionId}`;
+      paths[path] = {
+        post: {
+          operationId: actionId,
+          description: description,
+          parameters: [],
+          requestBody: {
+            description: `Input parameters for the action ${actionId} in BS`,
+            required: true,
+            content: {
+              'application/json;charset=utf-8': {
+                schema: {
+                  $ref: `#/components/schemas/${schemaName}`,
+                },
+              },
+            },
+          },
+          responses: {
+            default: {
+              description: `Action ${actionId} response`,
+              content: {
+                'application/json;charset=utf-8': {
```

Comment:
I believe you was going to drop the `charset=utf-8` bit?

Reply:
👍